### ISSUE TITLE
ERROR Unexpected exception: generator raised StopIteration

### ISSUE THREAD

**Issue by LoekvanKooten** [modernmt/modernmt#473]
=========== TRAINING STARTED ===========

ENGINE:   engdut
LANGUAGE: en > nl

(1/4) Cleaning corpora
    (1/6) Cleaning corpora...                                    DONE in 1m 4s
    (2/6) De-duplicate corpora...                                DONE in 4s
    (3/6) Preprocess corpora...                                  DONE in 36s
    (4/6) Creating aligner filter...                             DONE in 2m 23s
    (5/6) Scoring corpora...                                     DONE in 22s
    (6/6) Apply aligner-based filter...                          
ERROR Unexpected exception: generator raised StopIteration

**Comment by LoekvanKooten** [comment]
`2019-05-06 18:57:26,251 [INFO] - Training started: engine=engdut, lang_pair=en__nl
2019-05-06 18:57:26,274 [INFO] - Training step "clean" started
2019-05-06 18:57:26,275 [INFO] - Training step "clean" started
2019-05-06 18:57:26,275 [DEBUG] - java -DentityExpansionLimit=0 -DtotalEntitySizeLimit=0 -Djdk.xml.totalEntitySizeLimit=0 -Xmx14398m -cp /home/loek/mmt9/modernmt/build/mmt-4.0.jar -Dmmt.home=/home/loek/mmt9/modernmt -Djava.library.path=/home/loek/mmt9/modernmt/build/lib eu.modernmt.cli.CleaningPipelineMain -s en -t nl --input data/engdut --output /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/clean_corpora --output-format parallel
2019-05-06 18:58:30,679 [INFO] - Training step "clean" completed in 1m 4s
2019-05-06 18:58:30,679 [INFO] - Training step "dedup" started
2019-05-06 18:58:30,680 [DEBUG] - java -cp /home/loek/mmt9/modernmt/build/mmt-4.0.jar -Dmmt.home=/home/loek/mmt9/modernmt -Djava.library.path=/home/loek/mmt9/modernmt/build/lib eu.modernmt.cli.DeduplicationMain -s en -t nl --input /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/clean_corpora --output /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/dedup_corpora
2019-05-06 18:58:35,096 [INFO] - Training step "dedup" completed in 4s
2019-05-06 18:58:35,096 [INFO] - Training step "preprocess" started
2019-05-06 18:58:35,096 [DEBUG] - java -cp /home/loek/mmt9/modernmt/build/mmt-4.0.jar -Dmmt.home=/home/loek/mmt9/modernmt -Djava.library.path=/home/loek/mmt9/modernmt/build/lib eu.modernmt.cli.TrainingPipelineMain -s en -t nl --output /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/preprocessed_corpora --input /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/dedup_corpora
2019-05-06 18:59:10,942 [INFO] - Training step "preprocess" completed in 35s
2019-05-06 18:59:10,942 [INFO] - Training step "make_filter" started
2019-05-06 18:59:10,942 [DEBUG] - /home/loek/mmt9/modernmt/build/bin/fa_build -s en -t nl -i /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/preprocessed_corpora -m /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/aligner/en__nl.fam -I 4 --case-insensitive --no-favor-diagonal
Build started with options: {alpha=0.01, buffer_size=10000, case_sensitive=false, favor_diagonal=false, initial_diagonal_tension=4, iterations=4, max_length=80, optimize_tension=true, prob_align_null=0.08, pruning=1e-20, threads=8, use_null=true, variational_bayes=true, vocabulary_threshold=0.9999}
Creating vocabulary... DONE in 11.9056s
== Forward model training ==
Initial setup... DONE in 16.9181s
Iteration 1:
	Computing alignments... DONE in 9.49803s
	Normalizing translation table... DONE in 2.56821s
Iteration 2:
	Computing alignments... DONE in 9.30646s
	Normalizing translation table... DONE in 2.55583s
Iteration 3:
	Computing alignments... DONE in 9.28383s
	Normalizing translation table... DONE in 2.52201s
Iteration 4:
	Computing alignments... DONE in 9.27029s
	Normalizing translation table... DONE in 2.52185s
Pruning model... DONE in 0.562266s

Training DONE in 65.009s

== Backward model training ==
Initial setup... DONE in 15.9548s
Iteration 1:
	Computing alignments... DONE in 9.56978s
	Normalizing translation table... DONE in 2.56889s
Iteration 2:
	Computing alignments... DONE in 9.36464s
	Normalizing translation table... DONE in 2.56779s
Iteration 3:
	Computing alignments... DONE in 8.76275s
	Normalizing translation table... DONE in 2.55011s
Iteration 4:
	Computing alignments... DONE in 8.64442s
	Normalizing translation table... DONE in 2.57183s
Pruning model... DONE in 0.555362s

Training DONE in 63.1126s

Writing model...DONE in 0.827994s
2019-05-06 19:01:34,158 [INFO] - Training step "make_filter" completed in 2m 23s
2019-05-06 19:01:34,159 [INFO] - Training step "score" started
2019-05-06 19:01:34,159 [DEBUG] - /home/loek/mmt9/modernmt/build/bin/fa_score -s en -t nl -m /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/aligner/en__nl.fam -i /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/preprocessed_corpora -o /home/loek/mmt9/modernmt/runtime/engdut/tmp/training/_temp_cleaning/preprocessed_corpora
2019-05-06 19:01:56,708 [INFO] - Training step "score" completed in 22s
2019-05-06 19:01:56,708 [INFO] - Training step "apply_filter" started
2019-05-06 19:01:56,708 [INFO] - Applying aligner filter with: good_avg = -4.007930, good_std_dev = 1.447450, bad_avg = -12.526200, bad_std_dev = 3.490710`

**Comment by davidecaroselli** [comment]
Hi @LoekvanKooten 

just pushed the fix in `versions/v4.0` can you give it a try?

Thanks!

**Comment by LoekvanKooten** [comment]
@davidecaroselli 

Thank you! Please note that requirements.txt still needs to be changed into:

`regex
sacrebleu==1.3.1
requests
cachetools
torch
torchvision
fairseq==0.6.2`

**Comment by davidecaroselli** [comment]
Nope.

Unfortunately [pytorch](https://pytorch.org/) is not (yet) offering that shortcut. If you see on their website, with options Stable (1.1) > Linux > Pip > Python3.6 > 10.0 the command they require you to run is:
```bash
pip3 install https://download.pytorch.org/whl/cu100/torch-1.1.0-cp36-cp36m-linux_x86_64.whl
```
so unless they change this, I cannot use simple "torch" repo.  In your case I think there were problem because you used Python 3.7 that has a different pytorch wheel:
```bash
pip3 install https://download.pytorch.org/whl/cu100/torch-1.1.0-cp37-cp37m-linux_x86_64.whl
```

**Comment by LoekvanKooten** [comment]
That works!
